python_src/models/police_report.py

```diff
@@ -0,0 +1,251 @@
+from __future__ import annotations
+from typing import Tuple
+from abc import ABC, abstractmethod
+
+CAT_A = 'A'
+CAT_B = 'B'
+CAT_C = 'C'
+CAT_D = 'D'
+
+CAT_A_DATA = ["ANY"]
+CAT_B_DATA = ['F', 'M', 'B']
+CAT_C_DATA = ['L', 'R']
+CAT_D_DATA = ["FL", "FR", "ML", "MR", "BL", "BR"]
+
+CATEGORIES = [{"type": CAT_A, "data": CAT_A_DATA},
+              {"type": CAT_B, "data": CAT_B_DATA},
+              {"type": CAT_C, "data": CAT_C_DATA},
+              {"type": CAT_D, "data": CAT_D_DATA}]
+
+NO_CRASH = 0
+
+
+class Creator(ABC):
+    @abstractmethod
+    def factory_method(self):
+        pass
+
+    def match_operation(self, outputs: list, targets: list) -> Tuple[int, int, int]:
+        police_report = self.factory_method()
+        return police_report.process(outputs, targets)
+
+
+"""
+Concrete Creators override the factory method in order to change the resulting
+report's type.
+"""
+
+
+class ConcreteCreatorA(Creator):
+    def factory_method(self) -> Report:
+        return ReportTypeA()
+
+
+class ConcreteCreatorB(Creator):
+    def factory_method(self) -> Report:
+        return ReportTypeB()
+
+
+class ConcreteCreatorC(Creator):
+    def factory_method(self) -> Report:
+        return ReportTypeC()
+
+
+class ConcreteCreatorD(Creator):
+    def factory_method(self) -> Report:
+        return ReportTypeD()
+
+
+class Report(ABC):
+    """
+    The Report interface declares the operations that all concrete reports
+    must implement.
+    """
+
+    @abstractmethod
+    def process(self, outputs: list, targets: list) -> Tuple[int, int, int]:
+        """
+        Return the amount of matched crashed elements(components/sides/parts),
+               the amount of matched non-crashed elements(components/sides/parts)
+               the amount of "matchable" elements for a specific report
+        """
+        pass
+
+    @staticmethod
+    def _match(category: list,
+                           t1: list, o1: list, t2: list, o2: list) -> Tuple[int, int]:
+        """
+        Count the amount of matched crashed elements(components/sides/parts) and
+              the amount of matched non-crashed elements(components/sides/parts)
+
+        Args:
+            t1 (list): List of expected CRASHED element from a police report
+            o1 (list): List of expected CRASHED element from a simulation
+            t2 (list): List of expected NON-CRASHED element from a police report
+            o2 (list): List of expected NON-CRASHED element from a simulation
```

Comment:
What's the meaning of `t` and `o` here? Can we have more descriptive names? expected_crashes vs observed_crashes? chrases_from_police_report vs crashes_from_simulation?

Reply:
I renamed them to 
- `crashes_from_police_report`
- `crashes_from_simulation`
- `non_crashes_from_police_report`
- `non_crashes_from_simulation`